### PR TITLE
Rocos 2504 dynamic messaging json marshalling

### PR DESCRIPTION
**Pull request by hoani**
Closes https://rocosglobal.atlassian.net/browse/ROCOS-2504

Improves Marshalling/Unmarshalling performance.

TL;DR - Performance is improved by about 50%, but much more noticeable gains can be made by not using the `encoding/JSON` libraries `json.Marshal` function which will eat up the bulk of the performance gains we make by iterating over the marshalled data and validating it in `json.compact`

## Changes

* Broke all DynamicMessage JSON related functionality into a separate set of files
* Wrote a full marshaller for DynamicMessage data
* Wrote a whole load of marshaller/unmarshaller tests
* Rewrote unmarshal for both performance and reliability
  * Removed some bugs to do with strings not being unmarshalled with correct escapes
  * Arrays unmarshalling can now return error rather than silently not throw until some unexpected time later

## Performance

General benchmarks look ok; however, we get much better performance if we don't marshal through `encoding/JSON` and just call the `(*DynamicMessage).MarshalJSON` function directly. This would require some interface changes in the users of `DynamicMessage` and is not possible to do from this repository

Normalized rosgo Marshal Speed Improvements(1)
Normalized rosgo Unmarshal Speed Improvements(1)

Many of the performance improvements were gained on small messages. This will not have a big impact on the agent, because small message marshalling/unmarshalling does not represent a performance bottleneck in the rocos agent.

### Integration Performance Testing

For small packets, JSON marshalling is in the noise of the performance (roughly 0.1 to 1% of CPU); so the performance improvement was immeasurable.

For large arrays of bytes, the JSON marshalling with these changes are roughly equivalent in performance, and for other arrays the performance improvement only sits at around 2-5% of the overall CPU load.

For these reasons, the performance between this branch and master had uninteresting results.

However, by having the agent also omit using `encoding/json` we get the following results using a ros subscription to a point cloud with roughly 500kb of data per packet:

This branch with `encoding/JSON`:
image

This branch without `encoding/JSON`:
image

The ros node is clearly not a bottle neck of the rocos-agent at the moment; however, the ROS node becomes significantly lighter weight by dropping the `encding/json` checks.

Profile files:
[marshal-withcheck.pdf](https://github.com/team-rocos/rosgo/files/5924464/marshal-withcheck.pdf)
[marshal-nocheck.pdf](https://github.com/team-rocos/rosgo/files/5924466/marshal-nocheck.pdf)

### CODE REVIEW

File: ros/dynamic_message_json.go [hunk 1]
```diff
@@ -0,0 +1,1231 @@
...
+
+func unmarshalArray(msgType *DynamicMessageType, value []byte, field *libgengo.Field, dest *interface{}) error {
+	var err error
+	var unmarshalledLength int
+
+	size := field.ArrayLen
+	if size < 0 {
+		size = 0
+	}
+	switch field.GoType {
```
Comment (edwinhayes) [inline ros/dynamic_message_json.go:825]: Do you think that it would be worthwhile changing `field.GoType` from a string to an enum?  Lots string comparison in all these case statements.

**Review by edwinhayes — APPROVED** [review]
All looks good; just the thing I note below.  Thoughts on whether it would be a worthwhile exercise?

**Comment by edwinhayes** [comment]
Think this is a relevant issue to consider as part of this:

https://github.com/golang/go/issues/34701
https://philpearl.github.io/post/json_own_way/
https://github.com/wI2L/jettison

It raises a good point; in your new profiles, GC is a major component of CPU time now, and marshalling likely isn't helping.  Maybe want to consider switching to a different library, and implementing appending type marshalers for ROS messages.

File: ros/dynamic_message_json.go (hunk 1 above)
Comment (hoani) [reply ros/dynamic_message_json.go:825]: Yep, I actually benchmarked that!

The performance improvement goes from 0.3ns per lookup to 30ns per lookup. I don't see it as a major bottleneck, but it would have a positive impact

Comment (edwinhayes) [reply ros/dynamic_message_json.go:825]: How long do you think it will take; I assume it would affect a number of files, because it's going from the `libgengo` part?  If you think it's only a few hours, then probably worth doing now, otherwise we can just ticket it for future reference.

Comment (hoani) [reply ros/dynamic_message_json.go:825]: Yeah, probably only a few hours; happy to take care of it now :)

Comment (hoani) [reply ros/dynamic_message_json.go]: Now using enums. Added a field called `BuiltInType` instead of replacing `GoType` with an enum. This is because `GoType` also represents strings which are not built-ins, and I didn't want to break something else by accident.

**Review by edwinhayes — APPROVED** [review]
LGTM.
